### ISSUE TITLE
how about adding remote-endpoint in the client span for rpc (eg, openfeign) ?

### ISSUE THREAD

**Issue by yangyshdan** [spring-cloud/spring-cloud-sleuth#1887]
I am using sleuth-zipkin to setup a tracing solution with openfeign client. the version I am using as parent is : 
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.4.4</version>
        <relativePath/>

My envionrment is very complext, that it is highli possible that some of the applicaions do not get instrumented with sleuth-zipkin. So in order to still draw an the topology as comprehensive as possible, is it possible to add the remote-endpoint in the "client" span for open-feign? such as below 

     {   "traceId":"761fa7ad18d9172b",
        "parentId":"761fa7ad18d9172b",
        "id":"514fb6c1bcb6357f",
        "kind":"CLIENT",
        "name":"get",
        "timestamp":1616993053736129,
        "duration":1948,
        "localEndpoint":{
            "serviceName":"application-1"
        },
        **"remoteEndpoint":{
            "serviceName":"application-2"  # ip is also okay. Currently, the remote endpoint is not provided for client span.
        },**

        "tags":{
            "http.method":"GET",
            "http.path":"/echo"
        }}

By this way, even if the server span is not provided, with the client span, one can still get the topology calculated.

**Comment by jonatan-ivanov** [comment]
TL;DR: You can use a `HttpTracingCustomizer` to add `remoteServiceName` or a `HttpRequestParser` to add arbitrary tags that you want to use, see details, below.

HttpTracing

I think `localEndpoint` is not provided by Sleuth but by [Brave](https://github.com/openzipkin/brave). Also, it seems [`remoteEndpoint`](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/brave/src/main/java/brave/Span.java#L144) is deprecated in favor of [`remoteServiceName`](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/brave/src/main/java/brave/Span.java#L161) but I don't think Brave can provide that. In some cases, Sleuth does support `remoteServiceName` (search for it in the codebase) but not in the case of http clients like Feign, I'm not sure why, maybe @marcingrzejszczak does.

Let me walk you through on what happens under the hood so that you will see how to change the behavior:
- Sleuth wraps instrumented http clients; in case of Feign, there are multiple wrappers, one of them is [`TracingFeignClient `](https://github.com/spring-cloud/spring-cloud-sleuth/blob/5ab1c13f0cacd7d46b8528427c825f16fbfac216/spring-cloud-sleuth-instrumentation/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TracingFeignClient.java)
- You can see that it is using a `HttpClientHandler`; in case of Brave, the implementation is [`BraveHttpClientHandler`](https://github.com/spring-cloud/spring-cloud-sleuth/blob/5ab1c13f0cacd7d46b8528427c825f16fbfac216/spring-cloud-sleuth-brave/src/main/java/org/springframework/cloud/sleuth/brave/bridge/BraveHttpClientHandler.java)
- If you look into this, you will see that it is just delegating the work to [`brave.http.HttpClientHandler`](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/instrumentation/http/src/main/java/brave/http/HttpClientHandler.java)
- In Brave's handler you can see that it is setting the [`remoteServiceName`](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/instrumentation/http/src/main/java/brave/http/HttpClientHandler.java#L150-L153) that [comes from](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/instrumentation/http/src/main/java/brave/http/HttpClientHandler.java#L100) a [`HttpTracing`](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/instrumentation/http/src/main/java/brave/http/HttpTracing.java#L105-L107) instance
- So we "only" need to provide the name of the service to this class in Brave
- If you search in the Sleuth autoconfig for this, you should find that when it is created, [`HttpTracingCustomizer`s are applied](https://github.com/spring-cloud/spring-cloud-sleuth/blob/5ab1c13f0cacd7d46b8528427c825f16fbfac216/spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/brave/instrument/web/BraveHttpConfiguration.java#L117-L121)
- If you look into [`HttpTracingCustomizer`](https://github.com/openzipkin/brave/blob/master/instrumentation/http/src/main/java/brave/http/HttpTracingCustomizer.java), you will see that it gets a `HttpTracing.Builder` that [can set the server name](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/instrumentation/http/src/main/java/brave/http/HttpTracing.java#L326-L330) but the setter is package private so you can't set it here (I don't know why) :(
- Though `HttpTracing` [has a setter for this (`clientOf`)](https://github.com/openzipkin/brave/blob/5be287b91c3d18da9fc7a597d71b12b27be6043c/instrumentation/http/src/main/java/brave/http/HttpTracing.java#L114-L116)
- So you can create your own `HttpTracing` bean and set the remote server name
  [this is how Sleuth creates it](https://github.com/spring-cloud/spring-cloud-sleuth/blob/5ab1c13f0cacd7d46b8528427c825f16fbfac216/spring-cloud-sleuth-autoconfigure/src/main/java/org/springframework/cloud/sleuth/autoconfig/brave/instrument/web/BraveHttpConfiguration.java#L80-L123)

If I omit all of the logic, basically you should replace
`return builder.build();`
to
`return builder.build().clientOf("myServer");`

This is just a workaround and I don't think this will work for multiple Feign clients, I'm going to open a new issue for Brave to see why isn't this possible (package-private setter):
```java
@Bean
HttpTracingCustomizer httpTracingCustomizer() {
    return builder -> builder.serverName("myServer");
}
``` 
It might work if your `@Configuration` class is in the `brave.http` package but if anyone asks, I wasn't the one who said this. 😃

I'm also going to talk to @marcingrzejszczak to see what we can do: injecting the server name should be easy, what I'm worried about is having multiple Feign clients.

**Comment by jonatan-ivanov** [comment]
fyi: https://github.com/openzipkin/brave/issues/1296

**Comment by KYAW1256** [comment]
Leave a comment

**Comment by marcingrzejszczak** [comment]
We are not planning on adding any new features for Sleuth
